Support caching Clang invocations with -fprofile-use and -fprofile-instr-use.

michaelwoerister opened the pull request:
This PR adds support for caching Clang invocations with `-fprofile-use` and `-fprofile-instr-use`, something that will be required for supporting [PGOed Rust compiler builds](https://github.com/rust-lang/rust/issues/79562).

The PR *only* implements support for Clang. I have not had time to look into how GCC works (although from a superficial glance it looks like it might be harder to find out which files to hash exactly). Adding support for Rust should be straightforward though.

I added some unit tests and verified that things work locally, but found it hard to write end-to-end tests because of Clang's PGO setup which requires the external `llvm-profdata` tool to generate the `.profdata` file. Let me know if this is something that is required.

In https://github.com/mozilla/sccache/issues/941 I expressed the concern that hashing these rather big `.profdata` files would take a long time. In practice this does not seem to be true. In my local tests compiling LLVM with a full cache took about 50 seconds. Caching the file hashes (so that most of the hashing work could be skipped) only reduced the total build time by a couple of seconds or so.

r? @luser

luser commented on src/compiler/gcc.rs:
```diff
@@ -275,6 +283,11 @@ where
             Some(ProfileGenerate) => profile_generate = true,
+            Some(ProfileUse(_)) => {
+                if compiler_kind != CCompilerKind::Clang {
+                    cannot_cache!(arg.flag_str().unwrap(), "only Clang supported".into())
```
Just use a string literal here in `cannot_cache!`, since we use the strings as buckets to count non-cacheable compile requests and output them in the stats. You could use "-fprofile-use but not clang" or something like that.

luser commented on src/compiler/gcc.rs:
```diff
@@ -341,6 +354,20 @@ where
+                    path.push("default.profdata");
+                }
+
+                extra_hash_files.push(path);
```
This feels like it belongs more in the `match` above? I'm not sure why `ExtraHashFile` has logic down here, this match exists just to split arguments up into ones that get passed to the preprocessor or not.

luser reviewed (approved):
This LGTM. Just two small things I would change.

michaelwoerister commented:
Thanks for the review, @luser! I think I addressed your comments.

@glandium, would you maybe be able to spare some time for merging/reviewing this PR? The new functionality is a necessary prerequisite for [distributing PGOed versions of rustc](https://github.com/rust-lang/rust/issues/79442).

glandium commented on src/compiler/gcc.rs:
```diff
@@ -169,7 +176,7 @@ counted_array!(pub static ARGS: [ArgInfo<ArgData>; _] = [
     flag!("-fprofile-arcs", ProfileGenerate),
     flag!("-fprofile-generate", ProfileGenerate),
-    take_arg!("-fprofile-use", OsString, Concatenated, TooHard),
+    take_arg!("-fprofile-use", PathBuf, Concatenated('='), ProfileUse),
```
`-fprofile-use` can also be used without a value associated with it. It looks like that's true of `-fprofile-instr-use` too, and you don't handle those cases. There's also `-fno-profile-instr-use`, which you might as well add support for while you're here.

michaelwoerister replied:
It looks like `-fprofile-use` without argument will try to read from `./default.profdata` -- and that the implementation here already handles this correctly. I'll add a unit test for that.

glandium commented on src/compiler/gcc.rs:
```diff
@@ -275,6 +283,21 @@ where
             Some(ProfileGenerate) => profile_generate = true,
+            Some(ProfileUse(path)) => {
+                if compiler_kind != CCompilerKind::Clang {
+                    cannot_cache!("-fprofile(-instr)-use caching only supported for Clang");
```
no need for the `(-instr)` part.

glandium commented on src/compiler/gcc.rs:
```diff
@@ -275,6 +283,21 @@ where
             }
             Some(ProfileGenerate) => profile_generate = true,
+            Some(ProfileUse(path)) => {
+                if compiler_kind != CCompilerKind::Clang {
```
Another way you could do this, without adding the compiler kind, is to simply add -fprofile-use to the clang list of arguments (see `args::test_multi_search`), in which case it would probably be more appropriate to call this `ClangProfileUse` rather than `ProfileUse`.

michaelwoerister replied:
Ah, you can "override" arguments that way? That does sound like the better approach for this case.

glandium commented on src/compiler/msvc.rs:
```diff
@@ -587,7 +587,7 @@ pub fn parse_arguments(
                 Some(SplitDwarf) | Some(TestCoverage) | Some(Coverage) | Some(DoCompilation)
                 | Some(Language(_)) | Some(Output(_)) | Some(TooHardFlag) | Some(XClang(_))
-                | Some(TooHard(_)) => cannot_cache!(arg
+                | Some(TooHard(_)) | Some(ProfileUse(_)) => cannot_cache!(arg
```
it shouldn't be too hard to add support for clang-cl here.

michaelwoerister commented:
Thanks for the review, @glandium!

> Can you also rebase on top of #971 after it's merged? Thank you. 

Will do.

> (BTW, did you intend your user email not to contain a TLD?)

That was a halfhearted (and probably ineffective) attempt to keep my email address somewhat protected from spammers. It set up git like that years ago and it seems to have stuck `:/`

michaelwoerister commented:
OK, I think the version I just pushed addresses your comments, @glandium. And it does look cleaner than the previous version `:)`

Note: I added "support" for the `-fno-profile-*` flags by explicitly marking them as `TooHardFlag`. I can look into making them actually cacheable but I'll only do so if you think it's worth the trouble. 

I was a bit surprised to find that `GCC::parse_arguments()` seems to treat unknown arguments as always cacheable. Is that on purpose? It seems like a somewhat dangerous default.

michaelwoerister commented:
@glandium, do you have an estimate on when this can be merged? Is there anything I can do to help?

dancing-leaves commented:
Hi, would love to see these changes upstreamed, tracking this PR for a while. Thanks for the work, and lets rebase and merge it already. :slightly_smiling_face:

FilipAndersson245 commented:
@glandium Any idea of when/if this pull request can be merged?

mitchhentges reviewed (approved):
Thanks Michael, and sorry about the review delay.
Let's get this landed :sunglasses:

michaelwoerister commented:
Well, that was quick :) Thanks for the review!
